Adding custom `PublishArtifact` to `MavenPublication` fails

Attempting to add an artifact which is a custom implementation of `PublishArtifact` to a `MavenPublication` fails because at some point the internals attempt to cast this to a `PublishArtifactInternal`.

What I expect to happen is defined in the user-guide: https://docs.gradle.org/current/dsl/org.gradle.api.publish.maven.MavenPublication.html#org.gradle.api.publish.maven.MavenPublication:artifact(java.lang.Object)

To paraphrase:

> The artifact method can take a variety of input:
> 
> * A PublishArtifact instance. Extension and classifier values are taken from the wrapped instance.
> * ...

According to the documentation, any object implementing `PublishArtifact` should be allowed.  but effectively only internally created artifacts are allowed due to this bug.

My (pseudo) code:

```
class ShadowPublishArtifact implements PublishArtifact, TaskDependency {
    ...
}

void doStuff() {
    ShadowPublishArtifact shadowPublishArtifact = ...;
    shadowMavenPublication = publishingExtension.getPublications().create( "mavenShadowArtifacts", MavenPublication.class );
    shadowMavenPublication.artifact( shadowPublishArtifact )
}
```

What actually happens:

```
* What went wrong:
A problem occurred evaluating project ':real-jakarta'.
> Could not create an instance of type org.gradle.api.publish.maven.internal.artifact.PublishArtifactBasedMavenArtifact.
   > class org.hibernate.build.gradle.jakarta.shadow.ShadowPublishArtifact cannot be cast to class org.gradle.api.internal.artifacts.PublishArtifactInternal (org.hibernate.build.gradle.jakarta.shadow.ShadowPublishArtifact is in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader @33d8bfc8; org.gradle.api.internal.artifacts.PublishArtifactInternal is in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader @240f8312)

....

Caused by: java.lang.ClassCastException: class org.hibernate.build.gradle.jakarta.shadow.ShadowPublishArtifact cannot be cast to class org.gradle.api.internal.artifacts.PublishArtifactInternal (org.hibernate.build.gradle.jakarta.shadow.ShadowPublishArtifact is in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader @33d8bfc8; org.gradle.api.internal.artifacts.PublishArtifactInternal is in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader @240f8312)
	at org.gradle.api.publish.maven.internal.artifact.PublishArtifactBasedMavenArtifact.<init>(PublishArtifactBasedMavenArtifact.java:29)
	at org.gradle.api.publish.maven.internal.artifact.PublishArtifactBasedMavenArtifact_Decorated.<init>(Unknown Source)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at org.gradle.internal.instantiation.generator.AsmBackedClassGenerator$InvokeConstructorStrategy.newInstance(AsmBackedClassGenerator.java:1969)
	at org.gradle.internal.instantiation.generator.AbstractClassGenerator$GeneratedClassImpl$GeneratedConstructorImpl.newInstance(AbstractClassGenerator.java:468)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:64)
	... 180 more
```


My env:

```
------------------------------------------------------------
Gradle 6.8
------------------------------------------------------------

Build time:   2021-01-08 16:38:46 UTC
Revision:     b7e82460c5373e194fb478a998c4fcfe7da53a7e

Kotlin:       1.4.20
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          11.0.6 (AdoptOpenJDK 11.0.6+10)
OS:           Linux 5.10.8-100.fc32.x86_64 amd64

```

--- a/subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/PublishArtifactInternal.java
+++ b/subprojects/core-api/src/main/java/org/gradle/api/internal/artifacts/PublishArtifactInternal.java
@@ -19,4 +19,12 @@ import org.gradle.api.artifacts.PublishArtifact;
 
 public interface PublishArtifactInternal extends PublishArtifact {
     boolean shouldBePublished();
+
+    static boolean shouldBePublished(PublishArtifact artifact) {
+        if (artifact instanceof PublishArtifactInternal) {
+            return ((PublishArtifactInternal) artifact).shouldBePublished();
+        }
+        // This can happen for custom publish artifacts
+        return true;
+    }
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/artifacts/publish/DecoratingPublishArtifact.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/artifacts/publish/DecoratingPublishArtifact.java
@@ -102,10 +102,6 @@ public class DecoratingPublishArtifact extends AbstractPublishArtifact implement
 
     @Override
     public boolean shouldBePublished() {
-        if (publishArtifact instanceof PublishArtifactInternal) {
-            return ((PublishArtifactInternal) publishArtifact).shouldBePublished();
-        }
-        // This can happen for custom publish artifacts
-        return true;
+        return PublishArtifactInternal.shouldBePublished(publishArtifact);
     }
 }

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/LazyPublishArtifact.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/LazyPublishArtifact.java
@@ -18,7 +18,6 @@ package org.gradle.api.internal.artifacts.dsl;
 
 import org.gradle.api.InvalidUserDataException;
 import org.gradle.api.Task;
-import org.gradle.api.artifacts.PublishArtifact;
 import org.gradle.api.file.FileSystemLocation;
 import org.gradle.api.internal.artifacts.PublishArtifactInternal;
 import org.gradle.api.internal.artifacts.publish.ArchivePublishArtifact;
@@ -89,7 +88,7 @@ public class LazyPublishArtifact implements PublishArtifactInternal {
         return new Date();
     }
 
-    private PublishArtifact getDelegate() {
+    private PublishArtifactInternal getDelegate() {
         if (delegate == null) {
             Object value = provider.get();
             if (value instanceof FileSystemLocation) {
@@ -127,6 +126,6 @@ public class LazyPublishArtifact implements PublishArtifactInternal {
 
     @Override
     public boolean shouldBePublished() {
-        return ((PublishArtifactInternal) getDelegate()).shouldBePublished();
+        return getDelegate().shouldBePublished();
     }
 }

--- a/subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishBasicIntegTest.groovy
+++ b/subprojects/ivy/src/integTest/groovy/org/gradle/api/publish/ivy/IvyPublishBasicIntegTest.groovy
@@ -198,6 +198,75 @@ class IvyPublishBasicIntegTest extends AbstractIvyPublishIntegTest {
     }
 
     @ToBeFixedForConfigurationCache
+    def "can publish custom PublishArtifact"() {
+        given:
+        settingsFile << "rootProject.name = 'root'"
+        buildFile << """
+            apply plugin: 'ivy-publish'
+            apply plugin: 'java'
+            group = 'org.gradle.test'
+            version = '1.0'
+            def writeFileProvider = tasks.register("writeFile") {
+                doLast {
+                    try (FileOutputStream out = new FileOutputStream("customArtifact.jar")) {}
+                }
+            }
+            def customArtifact = new PublishArtifact() {
+                @Override
+                String getName() {
+                    return "customArtifact"
+                }
+                @Override
+                String getExtension() {
+                    return "jar"
+                }
+                @Override
+                String getType() {
+                    return "jar"
+                }
+                @Override
+                String getClassifier() {
+                    return null
+                }
+                @Override
+                File getFile() {
+                    return new File("customArtifact.jar")
+                }
+                @Override
+                Date getDate() {
+                    return new Date()
+                }
+                @Override
+                TaskDependency getBuildDependencies() {
+                    return new TaskDependency() {
+                        @Override
+                        Set<? extends Task> getDependencies(Task task) {
+                            return Collections.singleton(writeFileProvider.get())
+                        }
+                    }
+                }
+            }
+            publishing {
+                repositories {
+                    ivy { url "${ivyRepo.uri}" }
+                }
+                publications {
+                    ivy(IvyPublication) {
+                        artifact customArtifact
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds 'publish'
+
+        then:
+        def module = ivyRepo.module('org.gradle.test', 'root', '1.0')
+        module.assertPublished()
+    }
+
+    @ToBeFixedForConfigurationCache
     def "warns when trying to publish a transitive = false variant"() {
         given:
         def javaLibrary = javaLibrary(ivyRepo.module('group', 'root', '1.0'))

--- a/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/AbstractIvyArtifact.java
+++ b/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/AbstractIvyArtifact.java
@@ -19,7 +19,6 @@ package org.gradle.api.publish.ivy.internal.artifact;
 import com.google.common.base.Strings;
 import org.gradle.api.internal.tasks.AbstractTaskDependency;
 import org.gradle.api.internal.tasks.DefaultTaskDependency;
-import org.gradle.api.internal.tasks.TaskDependencyInternal;
 import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
 import org.gradle.api.publish.internal.PublicationArtifactInternal;
 import org.gradle.api.publish.ivy.IvyArtifact;
@@ -114,7 +113,7 @@ public abstract class AbstractIvyArtifact implements IvyArtifact, PublicationArt
         return allBuildDependencies;
     }
 
-    protected abstract TaskDependencyInternal getDefaultBuildDependencies();
+    protected abstract TaskDependency getDefaultBuildDependencies();
 
     @Override
     public String toString() {
@@ -124,7 +123,7 @@ public abstract class AbstractIvyArtifact implements IvyArtifact, PublicationArt
     private class CompositeTaskDependency extends AbstractTaskDependency {
         @Override
         public void visitDependencies(TaskDependencyResolveContext context) {
-            getDefaultBuildDependencies().visitDependencies(context);
+            context.add(getDefaultBuildDependencies());
             additionalBuildDependencies.visitDependencies(context);
         }
     }

--- a/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/ArchiveTaskBasedIvyArtifact.java
+++ b/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/ArchiveTaskBasedIvyArtifact.java
@@ -20,6 +20,7 @@ import com.google.common.collect.ImmutableSet;
 import org.gradle.api.internal.tasks.DefaultTaskDependency;
 import org.gradle.api.internal.tasks.TaskDependencyInternal;
 import org.gradle.api.publish.ivy.internal.publisher.IvyPublicationIdentity;
+import org.gradle.api.tasks.TaskDependency;
 import org.gradle.api.tasks.bundling.AbstractArchiveTask;
 
 import java.io.File;
@@ -61,7 +62,7 @@ public class ArchiveTaskBasedIvyArtifact extends AbstractIvyArtifact {
     }
 
     @Override
-    protected TaskDependencyInternal getDefaultBuildDependencies() {
+    protected TaskDependency getDefaultBuildDependencies() {
         return buildDependencies;
     }
 

--- a/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/DerivedIvyArtifact.java
+++ b/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/DerivedIvyArtifact.java
@@ -16,9 +16,9 @@
 
 package org.gradle.api.publish.ivy.internal.artifact;
 
-import org.gradle.api.internal.tasks.TaskDependencyInternal;
 import org.gradle.api.publish.internal.PublicationInternal;
 import org.gradle.api.publish.ivy.IvyArtifact;
+import org.gradle.api.tasks.TaskDependency;
 
 import java.io.File;
 
@@ -59,8 +59,8 @@ public class DerivedIvyArtifact extends AbstractIvyArtifact {
     }
 
     @Override
-    protected TaskDependencyInternal getDefaultBuildDependencies() {
-        return (TaskDependencyInternal) original.getBuildDependencies();
+    protected TaskDependency getDefaultBuildDependencies() {
+        return original.getBuildDependencies();
     }
 
     @Override

--- a/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/FileBasedIvyArtifact.java
+++ b/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/FileBasedIvyArtifact.java
@@ -19,6 +19,7 @@ package org.gradle.api.publish.ivy.internal.artifact;
 import com.google.common.io.Files;
 import org.gradle.api.internal.tasks.TaskDependencyInternal;
 import org.gradle.api.publish.ivy.internal.publisher.IvyPublicationIdentity;
+import org.gradle.api.tasks.TaskDependency;
 
 import java.io.File;
 
@@ -59,7 +60,7 @@ public class FileBasedIvyArtifact extends AbstractIvyArtifact {
     }
 
     @Override
-    protected TaskDependencyInternal getDefaultBuildDependencies() {
+    protected TaskDependency getDefaultBuildDependencies() {
         return TaskDependencyInternal.EMPTY;
     }
 

--- a/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/PublishArtifactBasedIvyArtifact.java
+++ b/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/PublishArtifactBasedIvyArtifact.java
@@ -16,17 +16,18 @@
 
 package org.gradle.api.publish.ivy.internal.artifact;
 
+import org.gradle.api.artifacts.PublishArtifact;
 import org.gradle.api.internal.artifacts.PublishArtifactInternal;
-import org.gradle.api.internal.tasks.TaskDependencyInternal;
 import org.gradle.api.publish.ivy.internal.publisher.IvyPublicationIdentity;
+import org.gradle.api.tasks.TaskDependency;
 
 import java.io.File;
 
 public class PublishArtifactBasedIvyArtifact extends AbstractIvyArtifact {
-    private final PublishArtifactInternal artifact;
+    private final PublishArtifact artifact;
     private final IvyPublicationIdentity identity;
 
-    public PublishArtifactBasedIvyArtifact(PublishArtifactInternal artifact, IvyPublicationIdentity identity) {
+    public PublishArtifactBasedIvyArtifact(PublishArtifact artifact, IvyPublicationIdentity identity) {
         this.artifact = artifact;
         this.identity = identity;
     }
@@ -57,8 +58,8 @@ public class PublishArtifactBasedIvyArtifact extends AbstractIvyArtifact {
     }
 
     @Override
-    protected TaskDependencyInternal getDefaultBuildDependencies() {
-        return (TaskDependencyInternal) artifact.getBuildDependencies();
+    protected TaskDependency getDefaultBuildDependencies() {
+        return artifact.getBuildDependencies();
     }
 
     @Override
@@ -68,6 +69,6 @@ public class PublishArtifactBasedIvyArtifact extends AbstractIvyArtifact {
 
     @Override
     public boolean shouldBePublished() {
-        return artifact.shouldBePublished();
+        return PublishArtifactInternal.shouldBePublished(artifact);
     }
 }

--- a/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/SingleOutputTaskIvyArtifact.java
+++ b/subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/artifact/SingleOutputTaskIvyArtifact.java
@@ -22,6 +22,7 @@ import org.gradle.api.internal.tasks.AbstractTaskDependency;
 import org.gradle.api.internal.tasks.TaskDependencyInternal;
 import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
 import org.gradle.api.publish.ivy.internal.publisher.IvyPublicationIdentity;
+import org.gradle.api.tasks.TaskDependency;
 import org.gradle.api.tasks.TaskProvider;
 
 import java.io.File;
@@ -70,7 +71,7 @@ public class SingleOutputTaskIvyArtifact extends AbstractIvyArtifact {
     }
 
     @Override
-    protected TaskDependencyInternal getDefaultBuildDependencies() {
+    protected TaskDependency getDefaultBuildDependencies() {
         return buildDependencies;
     }
 

--- a/subprojects/ivy/src/test/groovy/org/gradle/api/publish/ivy/internal/artifact/IvyArtifactNotationParserFactoryTest.groovy
+++ b/subprojects/ivy/src/test/groovy/org/gradle/api/publish/ivy/internal/artifact/IvyArtifactNotationParserFactoryTest.groovy
@@ -18,7 +18,7 @@ package org.gradle.api.publish.ivy.internal.artifact
 
 import com.google.common.collect.ImmutableSet
 import org.gradle.api.Task
-import org.gradle.api.internal.artifacts.PublishArtifactInternal
+import org.gradle.api.artifacts.PublishArtifact
 import org.gradle.api.internal.file.FileResolver
 import org.gradle.api.internal.tasks.DefaultTaskDependency
 import org.gradle.api.publish.ivy.IvyArtifact
@@ -34,7 +34,7 @@ public class IvyArtifactNotationParserFactoryTest extends AbstractProjectBuilder
     def fileNotationParser = Mock(NotationParser)
     def task = Mock(Task)
     def taskDependency = new DefaultTaskDependency(null, ImmutableSet.of(task))
-    def publishArtifact = Stub(PublishArtifactInternal) {
+    def publishArtifact = Stub(PublishArtifact) {
         getName() >> 'name'
         getExtension() >> 'extension'
         getType() >> 'type'

--- a/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy
+++ b/subprojects/language-native/src/test/groovy/org/gradle/language/plugins/NativeBasePluginTest.groovy
@@ -17,12 +17,12 @@
 package org.gradle.language.plugins
 
 import org.gradle.api.Task
+import org.gradle.api.artifacts.PublishArtifact
 import org.gradle.api.component.ComponentWithVariants
 import org.gradle.api.component.PublishableComponent
 import org.gradle.api.component.SoftwareComponent
 import org.gradle.api.file.FileCollection
 import org.gradle.api.internal.artifacts.DefaultModuleVersionIdentifier
-import org.gradle.api.internal.artifacts.PublishArtifactInternal
 import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal
 import org.gradle.api.internal.component.SoftwareComponentInternal
 import org.gradle.api.internal.component.UsageContext
@@ -412,7 +412,7 @@ class NativeBasePluginTest extends Specification {
 
     def "adds Maven publications for component with main publication"() {
         def usage1 = Stub(UsageContext)
-        def artifact1 = Stub(PublishArtifactInternal)
+        def artifact1 = Stub(PublishArtifact)
         artifact1.getFile() >> projectDir.file("artifact1")
         usage1.artifacts >> [artifact1]
         def variant1 = Stub(PublishableVariant)
@@ -421,7 +421,7 @@ class NativeBasePluginTest extends Specification {
         variant1.getCoordinates() >> new DefaultModuleVersionIdentifier("my.group", "test_app_debug", "1.2")
 
         def usage2 = Stub(UsageContext)
-        def artifact2 = Stub(PublishArtifactInternal)
+        def artifact2 = Stub(PublishArtifact)
         artifact2.getFile() >> projectDir.file("artifact1")
         usage2.artifacts >> [artifact2]
         def variant2 = Stub(PublishableVariant)

--- a/subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishBasicIntegTest.groovy
+++ b/subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishBasicIntegTest.groovy
@@ -16,6 +16,7 @@
 
 package org.gradle.api.publish.maven
 
+
 import org.gradle.integtests.fixtures.ToBeFixedForConfigurationCache
 import org.gradle.integtests.fixtures.publish.maven.AbstractMavenPublishIntegTest
 import org.gradle.test.fixtures.maven.MavenLocalRepository
@@ -307,6 +308,83 @@ class MavenPublishBasicIntegTest extends AbstractMavenPublishIntegTest {
         module.assertPublished()
         def module2 = mavenRepo2.module('org.gradle.test', 'root', '1.0')
         module2.assertPublished()
+    }
+
+    def "can publish custom PublishArtifact"() {
+        given:
+        settingsFile << "rootProject.name = 'root'"
+        buildFile << """
+            apply plugin: 'maven-publish'
+            apply plugin: 'java'
+
+            group = 'org.gradle.test'
+            version = '1.0'
+
+            def writeFileProvider = tasks.register("writeFile") {
+                doLast {
+                    try (FileOutputStream out = new FileOutputStream("customArtifact.jar")) {}
+                }
+            }
+
+            def customArtifact = new PublishArtifact() {
+                @Override
+                String getName() {
+                    return "customArtifact"
+                }
+
+                @Override
+                String getExtension() {
+                    return "jar"
+                }
+
+                @Override
+                String getType() {
+                    return "jar"
+                }
+
+                @Override
+                String getClassifier() {
+                    return null
+                }
+
+                @Override
+                File getFile() {
+                    return new File("customArtifact.jar")
+                }
+
+                @Override
+                Date getDate() {
+                    return new Date()
+                }
+
+                @Override
+                TaskDependency getBuildDependencies() {
+                    return new TaskDependency() {
+                        @Override
+                        Set<? extends Task> getDependencies(Task task) {
+                            return Collections.singleton(writeFileProvider.get())
+                        }
+                    }
+                }
+            }
+            publishing {
+                repositories {
+                    maven { url "${mavenRepo.uri}" }
+                }
+                publications {
+                    maven(MavenPublication) {
+                        artifact customArtifact
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds 'publish'
+
+        then:
+        def module = mavenRepo.module('org.gradle.test', 'root', '1.0')
+        module.assertPublished()
     }
 
     def "warns when trying to publish a transitive = false variant"() {

--- a/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/AbstractMavenArtifact.java
+++ b/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/AbstractMavenArtifact.java
@@ -19,7 +19,6 @@ package org.gradle.api.publish.maven.internal.artifact;
 import com.google.common.base.Strings;
 import org.gradle.api.internal.tasks.AbstractTaskDependency;
 import org.gradle.api.internal.tasks.DefaultTaskDependency;
-import org.gradle.api.internal.tasks.TaskDependencyInternal;
 import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
 import org.gradle.api.publish.internal.PublicationArtifactInternal;
 import org.gradle.api.publish.maven.MavenArtifact;
@@ -75,7 +74,7 @@ public abstract class AbstractMavenArtifact implements MavenArtifact, Publicatio
         return allBuildDependencies;
     }
 
-    protected abstract TaskDependencyInternal getDefaultBuildDependencies();
+    protected abstract TaskDependency getDefaultBuildDependencies();
 
     @Override
     public final String toString() {
@@ -86,7 +85,7 @@ public abstract class AbstractMavenArtifact implements MavenArtifact, Publicatio
 
         @Override
         public void visitDependencies(TaskDependencyResolveContext context) {
-            getDefaultBuildDependencies().visitDependencies(context);
+            context.add(getDefaultBuildDependencies());
             additionalBuildDependencies.visitDependencies(context);
         }
     }

--- a/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/PublishArtifactBasedMavenArtifact.java
+++ b/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/PublishArtifactBasedMavenArtifact.java
@@ -18,15 +18,15 @@ package org.gradle.api.publish.maven.internal.artifact;
 
 import org.gradle.api.artifacts.PublishArtifact;
 import org.gradle.api.internal.artifacts.PublishArtifactInternal;
-import org.gradle.api.internal.tasks.TaskDependencyInternal;
+import org.gradle.api.tasks.TaskDependency;
 
 import java.io.File;
 
 public class PublishArtifactBasedMavenArtifact extends AbstractMavenArtifact {
-    private final PublishArtifactInternal publishArtifact;
+    private final PublishArtifact publishArtifact;
 
     public PublishArtifactBasedMavenArtifact(PublishArtifact publishArtifact) {
-        this.publishArtifact = (PublishArtifactInternal) publishArtifact;
+        this.publishArtifact = publishArtifact;
     }
 
     @Override
@@ -45,12 +45,12 @@ public class PublishArtifactBasedMavenArtifact extends AbstractMavenArtifact {
     }
 
     @Override
-    protected TaskDependencyInternal getDefaultBuildDependencies() {
-        return (TaskDependencyInternal) publishArtifact.getBuildDependencies();
+    protected TaskDependency getDefaultBuildDependencies() {
+        return publishArtifact.getBuildDependencies();
     }
 
     @Override
     public boolean shouldBePublished() {
-        return publishArtifact.shouldBePublished();
+        return PublishArtifactInternal.shouldBePublished(publishArtifact);
     }
 }

--- a/subprojects/maven/src/test/groovy/org/gradle/api/publish/maven/internal/artifact/MavenArtifactNotationParserFactoryTest.groovy
+++ b/subprojects/maven/src/test/groovy/org/gradle/api/publish/maven/internal/artifact/MavenArtifactNotationParserFactoryTest.groovy
@@ -19,10 +19,10 @@ package org.gradle.api.publish.maven.internal.artifact
 
 import com.google.common.collect.ImmutableSet
 import org.gradle.api.Task
+import org.gradle.api.artifacts.PublishArtifact
 import org.gradle.api.file.FileCollection
 import org.gradle.api.file.RegularFile
 import org.gradle.api.internal.TaskOutputsInternal
-import org.gradle.api.internal.artifacts.PublishArtifactInternal
 import org.gradle.api.internal.file.FileResolver
 import org.gradle.api.internal.provider.ProviderInternal
 import org.gradle.api.internal.tasks.DefaultTaskDependency
@@ -42,7 +42,7 @@ class MavenArtifactNotationParserFactoryTest extends AbstractProjectBuilderSpec 
     def dependencies = ImmutableSet.of(task)
     def taskDependency = new DefaultTaskDependency(null, dependencies)
     def fileNotationParser = Mock(NotationParser)
-    def publishArtifact = Stub(PublishArtifactInternal) {
+    def publishArtifact = Stub(PublishArtifact) {
         getExtension() >> 'extension'
         getClassifier() >> 'classifier'
         getFile() >> new File('foo')

--- a/subprojects/maven/src/test/groovy/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublicationTest.groovy
+++ b/subprojects/maven/src/test/groovy/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublicationTest.groovy
@@ -24,13 +24,13 @@ import org.gradle.api.artifacts.ExcludeRule
 import org.gradle.api.artifacts.ModuleDependency
 import org.gradle.api.artifacts.ModuleVersionIdentifier
 import org.gradle.api.artifacts.ProjectDependency
+import org.gradle.api.artifacts.PublishArtifact
 import org.gradle.api.attributes.Category
 import org.gradle.api.component.ComponentWithVariants
 import org.gradle.api.file.FileCollection
 import org.gradle.api.internal.CollectionCallbackActionDecorator
 import org.gradle.api.internal.artifacts.DefaultModuleVersionIdentifier
 import org.gradle.api.internal.artifacts.DependencyManagementTestUtil
-import org.gradle.api.internal.artifacts.PublishArtifactInternal
 import org.gradle.api.internal.artifacts.ivyservice.projectmodule.ProjectDependencyPublicationResolver
 import org.gradle.api.internal.attributes.ImmutableAttributes
 import org.gradle.api.internal.component.SoftwareComponentInternal
@@ -183,7 +183,7 @@ class DefaultMavenPublicationTest extends Specification {
     def "artifacts are taken from added component"() {
         given:
         def publication = createPublication()
-        def artifact = Mock(PublishArtifactInternal)
+        def artifact = Mock(PublishArtifact)
         artifact.file >> artifactFile
         artifact.classifier >> ""
         artifact.extension >> "jar"
@@ -215,11 +215,11 @@ class DefaultMavenPublicationTest extends Specification {
     def "multiple usages of a component can provide the same artifact"() {
         given:
         def publication = createPublication()
-        def artifact1 = Mock(PublishArtifactInternal)
+        def artifact1 = Mock(PublishArtifact)
         artifact1.file >> artifactFile
         artifact1.classifier >> ""
         artifact1.extension >> "jar"
-        def artifact2 = Mock(PublishArtifactInternal)
+        def artifact2 = Mock(PublishArtifact)
         artifact2.file >> artifactFile
         artifact2.classifier >> ""
         artifact2.extension >> "jar"

--- a/subprojects/publish/src/main/java/org/gradle/api/publish/internal/metadata/ModuleMetadataSpecBuilder.java
+++ b/subprojects/publish/src/main/java/org/gradle/api/publish/internal/metadata/ModuleMetadataSpecBuilder.java
@@ -183,8 +183,7 @@ class ModuleMetadataSpecBuilder {
     }
 
     private boolean shouldNotBePublished(PublishArtifact artifact) {
-        return artifact instanceof PublishArtifactInternal
-            && !((PublishArtifactInternal) artifact).shouldBePublished();
+        return !PublishArtifactInternal.shouldBePublished(artifact);
     }
 
     private ModuleMetadataSpec.AvailableAt availableAt(ModuleVersionIdentifier coordinates, ModuleVersionIdentifier targetCoordinates) {

